src/mr.c

```diff
@@ -36,65 +72,168 @@ rpma_mr_read(struct ibv_qp *qp,
 /* public librpma API */
 
 /*
- * rpma_mr_reg -- use rpma_peer_mr_reg()
+ * rpma_mr_reg -- create a local memory registration object
  */
 int
 rpma_mr_reg(struct rpma_peer *peer, void *ptr, size_t size, int usage,
-		enum rpma_mr_plt plt, struct rpma_mr_local **mr)
+		enum rpma_mr_plt plt, struct rpma_mr_local **mr_ptr)
 {
-	return RPMA_E_NOSUPP;
+	if (peer == NULL || ptr == NULL || size == 0 || mr_ptr == NULL)
+		return RPMA_E_INVAL;
+
+	struct rpma_mr_local *mr;
+	mr = (struct rpma_mr_local *)Malloc(sizeof(struct rpma_mr_local));
+	if (mr == NULL)
+		return RPMA_E_NOMEM;
+
+	struct ibv_mr *ibv_mr;
+	int ret = rpma_peer_mr_reg(peer, &ibv_mr, ptr, size,
+			usage_to_access(usage));
+	if (ret)
+		goto err_free_rpma_mr_local;
+
+	mr->ibv_mr = ibv_mr;
+	mr->plt = plt;
+	*mr_ptr = mr;
+
+	return 0;
+
+err_free_rpma_mr_local:
+	Free(mr);
+
+	return ret;
 }
 
 /*
- * rpma_mr_dereg -- use ibv_dereg_mr()
+ * rpma_mr_dereg -- delete a local memory registration object
  */
 int
-rpma_mr_dereg(struct rpma_mr_local **mr)
+rpma_mr_dereg(struct rpma_mr_local **mr_ptr)
 {
-	return RPMA_E_NOSUPP;
+	if (mr_ptr == NULL)
+		return RPMA_E_INVAL;
+
+	if (*mr_ptr == NULL)
+		return 0;
+
+	int ret = 0;
+	struct rpma_mr_local *mr = *mr_ptr;
+	if (mr->ibv_mr) {
+		errno = ibv_dereg_mr(mr->ibv_mr);
+		if (errno) {
+			Rpma_provider_error = errno;
+			ret = RPMA_E_PROVIDER;
+		}
+	}
+
+	Free(mr);
+	*mr_ptr = NULL;
+
+	return ret;
 }
 
 /*
- * rpma_mr_serialize_get_size -- XXX
+ * rpma_mr_serialize_get_size -- size of the buffer for serialized memory region
  */
 size_t
 rpma_mr_serialize_get_size()
 {
-	return sizeof(struct rpma_mr_remote); /* XXX alignment? */
+	/*
+	 * Calculate a sum of sizes of rpma_mr_remote structure's fields:
+	 * sum = raddr + size + rkey + plt
+	 *
+	 * Note: plt has two possible values so it fits perfectly into a single
+	 * uint8_t value.
+	 */
+	return sizeof(uint64_t) + sizeof(uint64_t) + sizeof(uint32_t) +
+			sizeof(uint8_t);
 }
 
 /*
- * rpma_mr_serialize -- XXX
+ * rpma_mr_serialize -- serialize a memory region
  */
 int
 rpma_mr_serialize(struct rpma_mr_local *mr, char *buff)
```

Comment:
I'm not sure what this function does. It doesn't serialize any message, it simply writes a header to the buffer. Same for deserialization.